utils/parametric-components.stanza

```diff
@@ -0,0 +1,537 @@
+      val paraC = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val paraL = HashTable<Double|Toleranced, Vector<JITXObject>>()
+
+      ; Need to recursively retrieve all instances (including instances of modules.)
+      ; Loop runs twice: 1 find, 1 add part.
+      for i in list-all-instances() do :
+        if has-property?(i.parametric) :
+          ; println("Sorting part %_ with desired value %_" % [ref(i) property(i.resistance)])
+          if has-property?(i.resistor) :
+            match(get?(paraR, property(i.resistance))) : 
+              (v:Vector<JITXObject>) : add(v,i)
+              (_) : 
+                val t = Vector<JITXObject>()
+                add(t,i)
+                set(paraR, property(i.resistance), t)
+          else if has-property?(i.capacitor) :
+            match(get?(paraC, property(i.capacitance))) : 
+              (v:Vector<JITXObject>) : add(v,i)
+              (_) : 
+                val t = Vector<JITXObject>()
+                add(t,i)
+                set(paraC, property(i.capacitance), t)
+          else if has-property?(i.inductor) :
+            match(get?(paraL, property(i.inductance))) : 
+              (v:Vector<JITXObject>) : add(v,i)
+              (_) : 
+                val t = Vector<JITXObject>()
+                add(t,i)
+                set(paraL, property(i.inductance), t)
+          else :
+            fatal("Parametric component is not a resistor, capacitor, or inductor.")
+        else if has-property?(i.resistance) :
+          val r = property(i.resistance)
+          ; println("Found standard resistor (%_) of value (%_)" % [ref(i), r])
+          match(get?(knownR, r)) : 
+            (v:Vector<JITXObject>) : add(v,i)
+            (_) : 
+              val t = Vector<JITXObject>()
+              add(t,i)
+              set(knownR, r, t)
+        else if has-property?(i.capacitance) :
+          val c = property(i.capacitance)
+          ; println("Found standard capacitor (%_) of value (%_)" % [ref(i), r])
+          match(get?(knownC, c)) : 
+            (v:Vector<JITXObject>) : add(v,i)
+            (_) : 
+              val t = Vector<JITXObject>()
+              add(t,i)
+              set(knownC, c, t)
+        else if has-property?(i.inductance) :
+          val c = property(i.inductance)
+          ; println("Found standard inductor (%_) of value (%_)" % [ref(i), r])
+          match(get?(knownL, c)) : 
+            (v:Vector<JITXObject>) : add(v,i)
+            (_) : 
+              val t = Vector<JITXObject>()
+              add(t,i)
+              set(knownL, c, t)
+    
+      ; we have 2n Hashtables here 
+      ; paraR,paraC,.. are the parametric components
+      ; knownR,knownC,... are the known fixed components (non-parametric)
+      resolve-components(paraR, knownR)
+      resolve-components(paraC, knownC)
+      resolve-components(paraL, knownL)
+
+public defn list-all-instances () -> Vector<JITXObject> :
```

Comment:
In short
```
public defn list-all-instances () -> Seqable<JITXObject> :
  let loop (m:JITXObject = self) :
    for i in single-instances(m) seq-cat :
      cat([i], loop(i))
```

Reply:
Surprisingly you don't need the `inside pcb-module`.
I tried it with:
```
#use-added-syntax(jitx)
defpackage ocdb/designs/grid-resistors :
  import core

  import jitx
  import jitx/commands
  import ocdb/utils/defaults
  import ocdb/utils/generic-components

val board-shape = RoundedRectangle(10.0, 10.0, 0.5)

public defn list-all-instances () -> Seqable<JITXObject> :
  let loop (m:JITXObject = self) :
    for i in single-instances(m) seq-cat :
      cat([i], loop(i))

pcb-module main :
  inst i : database-part(["vendor_part_numbers.lcsc" => "C2290"])

  println $ to-tuple $ list-all-instances()

defn run-design (circuit: Instantiable) :
  make-default-board(circuit, 2, board-shape)

run-design(main)
```

```
Current working directory: "/Users/philippe/dev/jitx-client"
Loading JITX Shell...
[Instance(L294/D265(relative))]
```

Reply:
Nice